xarray/core/common.py

```diff
@@ -1765,7 +1789,38 @@ def _full_like_variable(other, fill_value, dtype: DTypeLike = None):
     return Variable(dims=other.dims, data=data, attrs=other.attrs)
 
 
-def zeros_like(other, dtype: DTypeLike = None):
+@overload
+def zeros_like(other: DataArray, dtype: DTypeLikeSave) -> DataArray:
+    ...
+
+
+@overload
+def zeros_like(other: Dataset, dtype: DTypeMaybeMapping) -> Dataset:
+    ...
+
+
+@overload
+def zeros_like(other: Variable, dtype: DTypeLikeSave) -> Variable:
+    ...
+
+
+@overload
+def zeros_like(
+    other: Dataset | DataArray, dtype: DTypeMaybeMapping = None
+) -> Dataset | DataArray:
```

Comment:
Do you know why this overload is required? I had thought that `DataArray` couldn't take a `DTypeMaybeMapping`? (and same principle for the next two)

Reply:
I tried to adjust these locally. An issue I hit is that so much of our typing is on `Dataset` rather than `T_Dataset` and changing one seems to require changing its dependencies, which makes it difficult to change gradually. 20 minutes into me trying to change this, I had 31 errors in 6 files!

To the extent you know which parts are Perfect vs Not-perfect-but-required-to-pass: If you want to add a comment for the ones the latter, that will make it easier for future travelers to know why things are as they are and hopefully change them.

Reply:
I was struggling a lot with `TypeVar` definitions.
Some other functions (like `polyval`) are calling `zeros_like` with `DataArray | Dataset`.

This means that a "simple" `TypeVar["T", DataArray, Dataset]` will not work.

I tried using `TypeVar["T", bound=DataArray|Dataset]` (recommended by some mypy people) but then the `if isinstance(x Dataset)` were causing problems (still not sure if that is a mypy bug or intended, my TypeVar knowledge is not good enough for that)...

So this was the only solution I could get to work.

Reply:
Additionally, the code will actually create a plain e.g. DataArray, so typevars with bounds are actually wrong here.

Reply:
Yeah. I think the proposed code is a big upgrade, and we can refine towards perfection in the future...

Reply:
FWIW I found [this SO answer](https://stackoverflow.com/a/59937840/3064736) helpful in clarifying the difference — i saw that I had upvoted it before — but I'm still not confident in how we should design these methods.